### PR TITLE
NoTrailingWhitespaceInStringFixer - introduction

### PR DESCRIPTION
**Pull request by gharlan**
This (risky) fixer removes trailing whitespace in strings.

The fixer is added to `@PhpCsFixer:risky`.

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -1218,6 +1218,12 @@ Choose from the list of available rules:
 
   There MUST be no trailing spaces inside comment or PHPDoc.
 
+* **no_trailing_whitespace_in_string** [@PhpCsFixer:risky]
+
+  There must be no trailing whitespace in strings.
+
+  *Risky rule: changing the the whitespaces in strings might affect string comparisons and outputs.*
```
Comment (julienfalque) [inline README.rst]: Extra "the".

File: src/Fixer/StringNotation/NoTrailingWhitespaceInStringFixer.php [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+    {
+        for ($index = $tokens->count() - 1, $last = true; $index >= 0; --$index, $last = false) {
+            $token = $tokens[$index];
+
+            if (!$token->isGivenKind([T_CONSTANT_ENCAPSED_STRING, T_ENCAPSED_AND_WHITESPACE, T_INLINE_HTML])) {
+                continue;
+            }
+
+            $isInlineHtml = $token->isGivenKind(T_INLINE_HTML);
+            $regex = $isInlineHtml && $last ? '/[\h]+(?=\R|$)/' : '/[\h]+(?=\R)/';
```
Comment (julienfalque) [inline src/Fixer/StringNotation/NoTrailingWhitespaceInStringFixer.php]:
```suggestion
            $regex = $isInlineHtml && $last ? '/\h+(?=\R|$)/' : '/\h+(?=\R)/';
```

File: src/Fixer/StringNotation/NoTrailingWhitespaceInStringFixer.php [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+
+            if (!$isInlineHtml || 0 === $index) {
+                $tokens[$index] = new Token([$token->getId(), $content]);
+
+                continue;
+            }
+
+            $prev = $index - 1;
+
+            if ($tokens[$prev]->equals([T_CLOSE_TAG, '?>'], false) && Preg::match('/^\R/', $content, $match)) {
```
Comment (SpacePossum) [inline src/Fixer/StringNotation/NoTrailingWhitespaceInStringFixer.php]: There seem to be no unit tests where this isn't true, please a case for it. I also think that the T_CLOSE_TAG `?>`doesn't need to be tested case insensitive.

**Comment by gharlan** [comment]
@kubawerlos thanks! test added, code fixed. ;)

**Comment by SpacePossum** [comment]
Thanks @gharlan  :+1:  nice to have this in!

**Pull request merged**